rust/src/snowball/snowball_env.rs

```diff
@@ -0,0 +1,404 @@
+        if self.cursor >= ket {
+            let new_cur = self.cursor as i32 + adjustment;
+            self.cursor = new_cur as usize;
+        } else if self.cursor > bra {
+            self.cursor = bra
+        }
+        self.current = Cow::from(result);
+        adjustment
+    }
+
+    /// Check if s is after cursor.
+    /// If so, move cursor to the end of s
+    pub fn eq_s(&mut self, s: &str) -> bool {
+        if self.cursor >= self.limit {
+            return false;
+        }
+        if self.current[self.cursor..].starts_with(s) {
+            self.cursor += s.len();
+            while !self.current.is_char_boundary(self.cursor) {
+                self.cursor += 1;
+            }
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Check if 's' is before cursor
+    /// If so, move cursor to the beginning of s
+    pub fn eq_s_b(&mut self, s: &str) -> bool {
+        if (self.cursor as i32 - self.limit_backward as i32) < s.len() as i32 {
+            false
+            // Check if cursor -s.len is a char boundry. if not well... return false obv
+        } else if !self.current.is_char_boundary(self.cursor - s.len()) ||
+                  !self.current[self.cursor - s.len()..].starts_with(s) {
+            false
+        } else {
+            self.cursor -= s.len();
+            true
+        }
+    }
+
+    /// Replace string between `bra` and `ket` with s
+    pub fn slice_from(&mut self, s: &str) -> bool {
+        let (bra, ket) = (self.bra, self.ket);
+        self.replace_s(bra, ket, s);
+        true
+    }
+
+    /// Move cursor to next charater
+    pub fn next_char(&mut self) {
+        self.cursor += 1;
+        while !self.current.is_char_boundary(self.cursor) {
+            self.cursor += 1;
+        }
+    }
+
+    /// Move cursor to previous character
+    pub fn previous_char(&mut self) {
+        self.cursor -= 1;
+        while !self.current.is_char_boundary(self.cursor) {
+            self.cursor -= 1;
+        }
+    }
+
+    pub fn byte_index_for_hop(&self, mut delta: i32) -> i32 {
+        if delta > 0 {
+            let mut res = self.cursor;
+            while delta > 0 {
+                res += 1;
+                delta -= 1;
+                while res <= self.current.len() && !self.current.is_char_boundary(res) {
+                    res += 1;
+                }
+            }
+            return res as i32;
+        } else if delta < 0 {
+            let mut res: i32 = self.cursor as i32;
+            while delta < 0 {
+                res -= 1;
+                delta += 1;
+                while res >= 0 && !self.current.is_char_boundary(res as usize) {
+                    res -= 1;
+                }
+            }
+            return res as i32;
+        } else {
+            return self.cursor as i32;
+        }
+    }
+
+    /// Check if the char the cursor points to is in the grouping
+    /// This is determined by weird magic stuff. I have no idea how it works
```

Comment:
Thank you for explaining. Is it ok if I copy this explanation into the code comments?